File: JAVA/PRACTICE/Java/QuickSort.java
package Java;

import java.util.Arrays;

class QuickSort{

    static void swap(int[]items, int leftPointer, int rightPointer){

        int temp;
        temp = items[leftPointer];
        items[leftPointer] = items[rightPointer];
        items[rightPointer] = temp;
    }

    static int partition(int[] items, int leftPointer, int rightPointer){
        
        int pivot = items[ (int)Math.floor( ( rightPointer + leftPointer)/2 )];

        while(leftPointer <= rightPointer){
        
        while(items[leftPointer] < pivot ){
            leftPointer ++;
        }

        while(items[rightPointer] > pivot){
            rightPointer --;
        }

        if(leftPointer <= rightPointer){
            swap(items, leftPointer, rightPointer);
            leftPointer ++;
            rightPointer --;
        }

        }

        return leftPointer;    

    }
   
    static int[] quickSort(int[] items, int leftPointer, int rightPointer){

        int index ; 
        
        if(items.length > 1){
            index = partition(items, leftPointer, rightPointer);
        if(leftPointer < index - 1){
            quickSort(items, leftPointer, index - 1);
        }
        if(index < rightPointer){
            quickSort(items, index, rightPointer);
        }
        }
        return items;
    }

    public static void main(String[] args) {
        
        int items[] = {9,12,4,1,3,8,5,10,6,11};

        int sortedItems[] = quickSort(items, 0, items.length - 1);

        System.out.println(Arrays.toString(sortedItems));
    }
}

File: JAVASCRIPT/PRACTICE/Js/QuickSort.js
const swap = (numbers, leftPointer, rightPointer) =>{
      
   [numbers[leftPointer], numbers[rightPointer]] = [numbers[rightPointer], numbers[leftPointer]];

};

const partition = (numbers, leftPointer, rightPointer) =>{

    let pivot = numbers[Math.floor((rightPointer + leftPointer)/2)];
    
    while( leftPointer <= rightPointer){

        while(numbers[leftPointer] < pivot){
           leftPointer ++;
        }

        while(numbers[rightPointer] > pivot){
            rightPointer --;
        }

        if(leftPointer <= rightPointer){
            swap(numbers,leftPointer,rightPointer);
            leftPointer ++;
            rightPointer --;
        }
    }
    return leftPointer;
};

const quickSort = (numbers, leftPointer, rightPointer) =>{

    let index;
    
    if(numbers.length > 1){
        index = partition(numbers,leftPointer, rightPointer);

        if(leftPointer < index -1){
            quickSort(numbers,leftPointer,index -1);
        }
        if(index < rightPointer){
            quickSort(numbers,index,rightPointer);
        }
    }
    return numbers;
};

const numbers = [5,19,20,50,21,2,18,4,1,49];

console.log(quickSort(numbers,0, numbers.length -1));